### DISCUSSION TITLE
How do you set the Background and settings?

### DISCUSSION THREAD

**Discussion by D1Ck3n** [realmazharhussain/gdm-settings#87] [Q&A]
Hello,

I am currently trying to configure GDM3 login for a project, but without using an external tool (like yours).

Could you explain me how you set the background and disable e.g. userlist?

I have found in the Arch wiki something about it and on Debian it also works, but Ubuntu or PopOS do not work.

i also tried to import the exported theme i created with your tool, but it doesn't work.

Would be cool from you if you could explain this to me.

Best
D1Ck3n

**Comment by realmazharhussain** [comment]
# Disclaimer
I'm just a human and I can make mistakes like everyone else. But we cannot through caution to the wind when dealing with such an import part of the OS. When fiddling with the Login Manager, a lot of unexpected things can happen and most of those will be breaking your installation. Take for example, issue #29 or #85.

Anything you do is your own responsibility.

# Instructions

The app mostly does things as told on the arch wiki. Basically, there are two types of changes the app makes.

1. Settings that are applied by modifying `/usr/share/gnome-shell/gnome-shell-theme.gresource`

   These include things like changing background, shell theme, and applying tweaks to the top bar

1. Settings that are applied through DConf

   These include most of the other settings

To elaborate, here it is

## DConf settings

Save the following content as `/etc/dconf/profile/gdm`

```
user-db:user
system-db:gdm
file-db:/usr/share/gdm/greeter-dconf-defaults
```

And save the following content after modifying it to your heart's content as `/etc/dconf/db/gdm.d/95-gdm-settings`

**Note:** You can also remove things from here that you don't care about. Also, you can the name of the file to anything you like. Just keep the number it starts with higher than any other file in that directory.

```ini
#-------- Interface ---------
[org/gnome/desktop/interface]
#----------------------------
cursor-theme='xcursor-breeze-snow'
icon-theme='Adwaita'
show-battery-percentage=true
clock-show-seconds=false
clock-show-weekday=true
clock-format='12h'
font-name='DejaVu Sans 10'
font-antialiasing='rgba'
font-hinting='slight'
text-scaling-factor=1.0

#-------- Sound ---------
[org/gnome/desktop/sound]
#------------------------
theme-name='Yaru'
event-sounds=true
input-feedback-sounds=false
allow-volume-above-100-percent=false

#-------------- Mouse ---------------
[org/gnome/desktop/peripherals/mouse]
#------------------------------------
accel-profile='default'
natural-scroll=false
speed=0.0

#-------------- Touchpad ---------------
[org/gnome/desktop/peripherals/touchpad]
#---------------------------------------
speed=0.35
tap-to-click=true
natural-scroll=true
two-finger-scrolling-enabled=true
disable-while-typing=true

#------------- Night Light --------------
[org/gnome/settings-daemon/plugins/color]
#----------------------------------------
night-light-enabled=true
night-light-temperature=uint32 3700
night-light-schedule-automatic=false
night-light-schedule-from=19.5
night-light-schedule-to=6.0

#----- Login Screen ----
[org/gnome/login-screen]
#-----------------------
logo='/usr/share/icons/manjaro/maia/logo_text_white_48px.png'
banner-message-enable=false
banner-message-text=''
disable-restart-buttons=false
disable-user-list=false
```

Run the command

```
sudo dconf update
```

After that, anytime you want to modify one of the settings present in `/etc/dconf/db/gdm/95-gdm-settings` (or whatever you named it), just edit the file and run the dconf command again.

## GResource Settings

These are a little complex. So, we'll need to do some bash scripting for this.

**Note:** This is a little bit dangerous. So, you should backup the `/usr/share/gnome-shell/gnome-shell-theme.gresource` before trying any of this.

Create a shell script with following content and save it as `extract-shell-theme.sh`

```bash
#!/bin/bash

shell_dir=extracted-theme/gnome-shell
gresource_file=/usr/share/gnome-shell/gnome-shell-theme.gresource

while read resource
do

  resource_name=${resource#/org/gnome/shell/theme/}
  file_path=${shell_dir}/${resource_name}
  file_dir=$(dirname "${file_path}")

  mkdir -p "${file_dir}"

  gresource extract ${gresource_file} "${resource}" > "${file_path}"

done <<< $(gresource list ${gresource_file})
```

Create another shell script with following content and save it as `compile-shell-theme.sh`

```bash
#!/bin/bash

theme_dir=${1:-extracted-theme}
shell_dir=${theme_dir}/gnome-shell
gresource_xml=/tmp/gnome-shell-theme.gresource.xml

cp "${shell_dir}"/gnome-shell.css "${shell_dir}"/gdm.css

echo -n '<?xml version="1.0" encoding="UTF-8"?>
<gresources>
 <gresource prefix="/org/gnome/shell/theme">
' > "${gresource_xml}"

while read file_path
do
  resource_path=${file_path#"${shell_dir}/"}
  echo "  <file>${resource_path}</file>" >> "${gresource_xml}"
done <<< $(find "${shell_dir}" -type f)

echo -n ' </gresource>
</gresources>
' >> "${gresource_xml}"

glib-compile-resources --sourcedir="${shell_dir}" --target=gnome-shell-theme.gresource "${gresource_xml}"
```

Now you can

1. Extract default shell theme by running `extract-shell-theme.sh`
2. Make changes to it
3. Compile it by running `compile-shell-theme.sh`
4. Replace the previous system-wide shell theme `/usr/share/gnome-shell/gnome-shell-theme.gresource` with the changed one using the following command.

   ```bash
   sudo cp gnome-shell-theme.gresource /usr/share/gnome-shell/gnome-shell-theme.gresource
   ```

   **Note:** It may seem like using `mv` instead of `cp` makes more sense here but if your distro has SELinux enabled, using `mv` can cause GNOME's white screen of death.
5. If and only if you are on Ubuntu or any derivative of Ubuntu, run the following command;

   ```bash
   sudo update-alternatives --set gdm-theme.gresource /usr/share/gnome-shell/gnome-shell-theme.gresource
   ```

   **Note:** You need to run this command only just one time but running it multiple times is also safe.

### Making Changes to the shell theme
Note that everything in this section is supposed to happen as step 2 of above instructions.

#### Change GDM Background Image
Save your background image as `extracted-theme/gnome-shell/background` and append following content at the end of `extracted-theme/gnome-shell/gnome-shell.css`

```css
#lockDialogGroup {
  background-image: url('resource:///org/gnome/shell/theme/background');
  background-size: cover;
}
```

#### Change GDM Background Color
Append following content at the end of `extracted-theme/gnome-shell/gnome-shell.css`

```css
#lockDialogGroup {
  background-color: blue;
}
```

Just replace `blue` with any color you like.

#### Change Text and Background Color of Top Bar
Append following content at the end of `extracted-theme/gnome-shell/gnome-shell.css`

```css
#panel .panel-button, #panel.login-screen .panel-button, #panel.unlock-screen .panel-button {
  color: green;
  background-color: red;
}
```

Just replace color names with whatever you like.

#### Disable Rounded Corners of Top Bar
Append following content at the end of `extracted-theme/gnome-shell/gnome-shell.css`

```css
#panel .panel-corner, #panel.login-screen .panel-corner, #panel.unlock-screen .panel-corner {
  -panel-corner-opacity: 0;
}
```

#### Disable Arrows of Top Bar
Append following content at the end of `extracted-theme/gnome-shell/gnome-shell.css`

```css
#panel .popup-menu-arrow, #panel.login-screen .popup-menu-arrow, #panel.unlock-screen .popup-menu-arrow {
  width: 0;
}
```

**Comment by realmazharhussain** [comment]
> i also tried to import the exported theme i created with your tool, but it doesn't work.

What do you mean? Does it not show up in Gnome Tweaks? What happens? A little more information, please.
